File: libs/east-node/packages/east-node-cli/src/snapshot.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

/*
 * Snapshot writer/reader — matches the cross-runtime format defined in
 * docs/snapshot-format.md. A .east-snapshot is an uncompressed POSIX ustar
 * archive containing manifest.json, ir.<ext>, and input-<N>.<ext>.
 */

import { createReadStream, createWriteStream, mkdtempSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { basename, extname, join } from 'node:path';
import { pipeline } from 'node:stream/promises';
import {
    StructType,
    ArrayType,
    StringType,
    IntegerType,
    DateTimeType,
    encodeJSONFor,
    decodeJSONFor,
    type ValueTypeOf,
} from '@elaraai/east';
import tar from 'tar-stream';

const SNAPSHOT_FORMAT_VERSION = 1n;

/* -------------------------------------------------------------------------- */
/*  Manifest schema — byte-identical across all runtime CLIs                   */
/* -------------------------------------------------------------------------- */

const SnapshotManifestType = StructType({
    version:    IntegerType,
    created_at: DateTimeType,
    runtime:    StructType({
        impl: StringType,
        cli:  StringType,
    }),
    ir:       StringType,
    inputs:   ArrayType(StringType),
    packages: ArrayType(StringType),
});

type SnapshotManifest = ValueTypeOf<typeof SnapshotManifestType>;

/* -------------------------------------------------------------------------- */
/*  Helpers                                                                    */
/* -------------------------------------------------------------------------- */

function extOf(path: string): string {
    const e = extname(basename(path));
    return e.startsWith('.') ? e.slice(1) : e;
}

/* -------------------------------------------------------------------------- */
/*  Writer                                                                     */
/* -------------------------------------------------------------------------- */

export async function writeSnapshot(params: {
    outPath:    string;
    irPath:     string;
    inputPaths: string[];
    packages:   string[];
    cliVersion: string;
}): Promise<void> {
    const { outPath, irPath, inputPaths, packages, cliVersion } = params;

    const irArchiveName = `ir.${extOf(irPath)}`;
    const inputArchiveNames = inputPaths.map((p, i) => `input-${i}.${extOf(p)}`);

    const manifest: SnapshotManifest = {
        version:    SNAPSHOT_FORMAT_VERSION,
        created_at: new Date(),
        runtime:    { impl: 'east-node', cli: cliVersion },
        ir:         irArchiveName,
        inputs:     inputArchiveNames,
        packages:   packages,
    };

    const manifestBytes = encodeJSONFor(SnapshotManifestType)(manifest);

    const pack = tar.pack();
    const out = createWriteStream(outPath);
    const done = pipeline(pack, out);

    await new Promise<void>((resolve, reject) => {
        const buf = Buffer.from(manifestBytes);
        pack.entry({ name: 'manifest.json', size: buf.length }, buf, err =>
            err ? reject(err) : resolve());
    });

    const irBytes = readFileSync(irPath);
    await new Promise<void>((resolve, reject) => {
        pack.entry({ name: irArchiveName, size: irBytes.length }, irBytes, err =>
            err ? reject(err) : resolve());
    });

    for (let i = 0; i < inputPaths.length; i++) {
        const bytes = readFileSync(inputPaths[i]!);
        const name = inputArchiveNames[i]!;
        await new Promise<void>((resolve, reject) => {
            pack.entry({ name, size: bytes.length }, bytes, err =>
                err ? reject(err) : resolve());
        });
    }

    pack.finalize();
    await done;
}

/* -------------------------------------------------------------------------- */
/*  Reader                                                                     */
/* -------------------------------------------------------------------------- */

export type SnapshotExtract = {
    irPath:     string;
    inputPaths: string[];
    packages:   string[];
    cleanup(): void;
};

export async function readSnapshot(inPath: string): Promise<SnapshotExtract> {
    const extract = tar.extract();
    const entries: Array<{ name: string; data: Buffer }> = [];

    const read = new Promise<void>((resolve, reject) => {
        extract.on('entry', (header, stream, next) => {
            const chunks: Buffer[] = [];
            stream.on('data', (c: Buffer) => chunks.push(c));
            stream.on('end', () => {
                entries.push({ name: header.name, data: Buffer.concat(chunks) });
                next();
            });
            stream.on('error', reject);
            stream.resume();
        });
        extract.on('finish', resolve);
        extract.on('error', reject);
    });

    await pipeline(createReadStream(inPath), extract);
    await read;

    const manifestEntry = entries.find(e => e.name === 'manifest.json');
    if (!manifestEntry) throw new Error(`Snapshot is missing manifest.json: ${inPath}`);

    const manifest: SnapshotManifest = decodeJSONFor(SnapshotManifestType)(manifestEntry.data);

    if (manifest.version !== SNAPSHOT_FORMAT_VERSION) {
        throw new Error(
            `Snapshot format version ${manifest.version} is not supported (expected ${SNAPSHOT_FORMAT_VERSION})`,
        );
    }

    const dir = mkdtempSync(join(tmpdir(), 'east-snapshot-'));
    for (const entry of entries) {
        if (entry.name === 'manifest.json') continue;
        writeFileSync(join(dir, entry.name), entry.data);
    }

    const irPath = join(dir, manifest.ir);
    const inputPaths = manifest.inputs.map(n => join(dir, n));

    return {
        irPath,
        inputPaths,
        packages: manifest.packages,
        cleanup: () => rmSync(dir, { recursive: true, force: true }),
    };
}

File: libs/east-py/packages/east-py-cli/east_py_cli/snapshot.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""Snapshot writer/reader — matches the cross-runtime format in
docs/snapshot-format.md.

A .east-snapshot is an uncompressed POSIX ustar archive containing
manifest.json, ir.<ext>, and input-<N>.<ext>. The manifest is east-JSON of
SnapshotManifestType (see the spec)."""

from __future__ import annotations

import io
import shutil
import tarfile
import tempfile
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from east.serialization.json import decode_json_for, encode_json_for
from east.types.types import (
    ArrayType,
    DateTimeType,
    IntegerType,
    StringType,
    StructType,
)

SNAPSHOT_FORMAT_VERSION = 1


SnapshotManifestType = StructType(
    [
        ("version", IntegerType),
        ("created_at", DateTimeType),
        ("runtime", StructType([("impl", StringType), ("cli", StringType)])),
        ("ir", StringType),
        ("inputs", ArrayType(StringType)),
        ("packages", ArrayType(StringType)),
    ]
)


@dataclass
class SnapshotExtract:
    """Paths to an extracted snapshot; call ``cleanup()`` to release."""

    ir_path: Path
    input_paths: list[Path]
    packages: list[str]
    _extract_dir: Path

    def cleanup(self) -> None:
        shutil.rmtree(self._extract_dir, ignore_errors=True)


def _ext_of(path: Path) -> str:
    s = path.suffix
    return s[1:] if s.startswith(".") else s


def _add_bytes(tar: tarfile.TarFile, name: str, data: bytes) -> None:
    info = tarfile.TarInfo(name=name)
    info.size = len(data)
    tar.addfile(info, io.BytesIO(data))


def write_snapshot(
    out_path: Path,
    ir_path: Path,
    input_paths: list[Path],
    packages: list[str],
    cli_version: str,
) -> None:
    """Write a .east-snapshot bundle. Must run before program execution so a
    crash still leaves the bundle behind."""
    ir_archive_name = f"ir.{_ext_of(ir_path)}"
    input_archive_names = [f"input-{i}.{_ext_of(p)}" for i, p in enumerate(input_paths)]

    manifest = {
        "version": SNAPSHOT_FORMAT_VERSION,
        "created_at": datetime.now(UTC),
        "runtime": {"impl": "east-py", "cli": cli_version},
        "ir": ir_archive_name,
        "inputs": input_archive_names,
        "packages": list(packages),
    }
    manifest_bytes = encode_json_for(SnapshotManifestType)(manifest)

    with tarfile.open(out_path, mode="w", format=tarfile.USTAR_FORMAT) as tar:
        _add_bytes(tar, "manifest.json", manifest_bytes)
        _add_bytes(tar, ir_archive_name, Path(ir_path).read_bytes())
        for name, p in zip(input_archive_names, input_paths, strict=True):
            _add_bytes(tar, name, Path(p).read_bytes())


def read_snapshot(in_path: Path) -> SnapshotExtract:
    """Extract a .east-snapshot bundle into a fresh temp dir and return the
    paths the CLI should use. Caller must ``cleanup()`` when done."""
    with tarfile.open(in_path, mode="r") as tar:
        manifest_member = tar.getmember("manifest.json")
        f = tar.extractfile(manifest_member)
        if f is None:
            raise ValueError(f"snapshot is missing manifest.json: {in_path}")
        manifest_bytes = f.read()

        manifest = decode_json_for(SnapshotManifestType)(manifest_bytes)

        if manifest["version"] != SNAPSHOT_FORMAT_VERSION:
            raise ValueError(
                f"snapshot format version {manifest['version']} is not supported "
                f"(expected {SNAPSHOT_FORMAT_VERSION})"
            )

        extract_dir = Path(tempfile.mkdtemp(prefix="east-snapshot-"))
        for member in tar.getmembers():
            if member.name == "manifest.json":
                continue
            tar.extract(member, path=extract_dir, filter="data")

    ir_path = extract_dir / manifest["ir"]
    input_paths = [extract_dir / name for name in manifest["inputs"]]

    return SnapshotExtract(
        ir_path=ir_path,
        input_paths=input_paths,
        packages=list(manifest["packages"]),
        _extract_dir=extract_dir,
    )
